src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -854,18 +864,19 @@ int MPIDI_OFI_mpi_finalize_hook(void)
         MPIDI_OFI_PROGRESS(0);
     MPIR_Assert(MPL_atomic_load_int(&MPIDI_OFI_global.am_inflight_inject_emus) == 0);
 
-    /* Tearing down endpoints */
-    for (i = 1; i < MPIDI_OFI_global.num_vnis; i++) {
-        mpi_errno = destroy_vni_context(i);
-        MPIR_ERR_CHECK(mpi_errno);
+    /* Tearing down endpoints in reverse order they were created */
+    for (i = MPIDI_OFI_global.num_vnis - 1; i >= 0; i--) {
+        for (int j = MPIDI_OFI_global.num_nics - 1; j >= 0; j--) {
+            mpi_errno = destroy_vni_context(i, j);
+            MPIR_ERR_CHECK(mpi_errno);
+        }
```

Comment:
My original thinking in separating vni `0` from the rest of the vnis was to differentiate vni 0. All other vnis are created equal (and destroyed equal, there is no ordering requirement, for example), but vni `0` can be used for bootstrapping or used to implement any amount of fancy schemes. For example, if we want to extend to have a dynamic vni connection, we can do that with a very minimal code change. The code for vni stuff is completely *orthogonal* to rest of the init/finalize. Now your code made vni 0 equal to all other vnis. On one hand, it seems simplifies the vni, but on the hand, you semantically made every vni as complicated as vni 0.

That's just some background information, I am not going to use it to block. But the ordering is not significant here, right? Unless there is specific reason, which you should comment, we should not do this reverse ordering, since it misleads developers to guess something may not be there. But if the ordering matters at all, I really would like to split vni `0-0` out and remove this ordering.

Reply:
I think the ordering does still matter if you're using scalable endpoints where the first VNI has more information than the rest and therefore needs to be kept around longest. If you're only using the domain mode, I think you're right.